File: leet/0091-decode-ways/decode-ways.go
func numDecodings(s string) int {
    if len(s) > 0 && s[0] == '0' {return 0}
    if len(s) < 2 {return 1}
    
    if (s[0] > '2') {
        return numDecodings(s[1:])
    } else {        
        if (s[0] == '1'){
            return numDecodings(s[1:]) + numDecodings(s[2:])
        } else {
            if (s[1] > '6') {return numDecodings(s[1:])}
            return numDecodings(s[1:]) + numDecodings(s[2:])
        }
    }
}

File: leet/0091-decode-ways/decode-ways.js
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
    let m = s.length
    let mp = new Map()

    let dp = function(n){
        if (mp.has(n)) return mp.get(n)

        let ret = 0
        if (n >= 0) {
            let i = parseInt(s[n])
            if (i > 0 && i < 27) ret += (n > 0 ? dp(n - 1) : 1)
            if (n > 0 && s[n - 1] != 0){
                let ii = parseInt(s[n - 1] + s[n])
                if(ii > 0 && ii < 27) ret += (n > 1 ? dp(n - 2) : 1)
            }
        }

        mp.set(n, ret)
        return ret
    }

    for (let i = 0; i < s.length; i++) dp(i)
    return mp.get(m - 1)
};
